vesc_driver/src/vesc_interface.cpp

```diff
@@ -50,49 +50,70 @@ class VescInterface::Impl
+  void packet_creation_thread();
+  void on_configure();
+  void connect(const std::string & port);
 
-  std::unique_ptr<std::thread> rx_thread_;
-  bool rx_thread_run_;
+  bool packet_thread_run_;
+  std::unique_ptr<std::thread> packet_thread_;
   PacketHandlerFunction packet_handler_;
   ErrorHandlerFunction error_handler_;
-  boost::asio::io_service io_service_;
-  boost::asio::serial_port serial_port_;
-  std::mutex serial_mutex_;
+  std::unique_ptr<drivers::serial_driver::SerialPortConfig> device_config_;
+  std::mutex buffer_mutex_;
+  std::string device_name_;
+  std::unique_ptr<IoContext> owned_ctx{};
+  std::unique_ptr<drivers::serial_driver::SerialDriver> serial_driver_;
+  
+  ~Impl(){
+    if (owned_ctx) {
+      owned_ctx->waitForExit();
+  }
+  }
+private:
+  std::vector<uint8_t> buffer_;
 };
 
-void VescInterface::Impl::rxThread()
+void VescInterface::Impl::serial_receive_callback(const std::vector<uint8_t> & buffer)
 {
-  Buffer buffer;
-  buffer.reserve(4096);
+  buffer_mutex_.lock(); // wait untill the current buffer read finishes
+  buffer_.reserve(buffer_.size() + buffer.size());
+  buffer_.insert(buffer_.end(), buffer.begin(), buffer.end());
+  buffer_mutex_.unlock();
+}
 
-  while (rx_thread_run_) {
+void VescInterface::Impl::packet_creation_thread()
+{
+  while (packet_thread_run_)
+  {
+    buffer_mutex_.lock();
     int bytes_needed = VescFrame::VESC_MIN_FRAME_SIZE;
-    if (!buffer.empty()) {
+    if (!buffer_.empty()){
       // search buffer for valid packet(s)
-      Buffer::iterator iter(buffer.begin());
-      Buffer::iterator iter_begin(buffer.begin());
-      while (iter != buffer.end()) {
+      auto iter = buffer_.begin();
+      auto iter_begin = buffer_.begin();
+      while (iter != buffer_.end())
+      {
         // check if valid start-of-frame character
         if (VescFrame::VESC_SOF_VAL_SMALL_FRAME == *iter ||
           VescFrame::VESC_SOF_VAL_LARGE_FRAME == *iter)
         {
           // good start, now attempt to create packet
           std::string error;
           VescPacketConstPtr packet =
-            VescPacketFactory::createPacket(iter, buffer.end(), &bytes_needed, &error);
+            VescPacketFactory::createPacket(iter, buffer_.end(), &bytes_needed, &error);
           if (packet) {
             // good packet, check if we skipped any data
-            if (std::distance(iter_begin, iter) > 0) {
-              std::ostringstream ss;
-              ss << "Out-of-sync with VESC, unknown data leading valid frame. Discarding " <<
-                std::distance(iter_begin, iter) << " bytes.";
-              error_handler_(ss.str());
-            }
+            //if (std::distance(iter_begin, iter) > 0) {
+              //std::ostringstream ss;
+              //ss << "Out-of-sync with VESC, unknown data leading valid frame. Discarding " <<
+              //  std::distance(iter_begin, iter) << " bytes.";
+              //error_handler_(ss.str());
+            //}
```

Comment:
Is there a reason this is commented out? If there is, it should be removed. If there is not, it should be un-commented.

Reply:
Hi Josh, we are working on a new version that is comparable with the old error handler. It was commented out because we cannot detect out of sync error with async receive method. We will have the revisions ready this weekend.

Reply:
Thanks! If you push the changes to this branch, the PR will automatically update. Looking forward to the updates!